### ISSUE TITLE
HMR on non-web react project

### ISSUE THREAD

**Issue by konsumer** [alangpierce/sucrase#436]
Maybe I am going about this totally wrong, but I am trying to use HMR with a cli app that uses [ink](https://github.com/vadimdemedes/ink). I couldn't seem to get it working without webpack, so I thought I would try out sucrase (much faster, btw, love it.)

It's still not hot-reloading, but I'm not sure what I need to do. Any suggestions would be welcome.

I have a branch of my demo-project [here](https://github.com/konsumer/ink-demo/tree/sucrase).

**Comment by alangpierce** [comment]
@konsumer Interesting use case! I've never tried using HMR like this. Unfortunately Sucrase itself probably won't be much help here, since at least the core of Sucrase is really just a way to compile an individual file, nothing stronger than what Babel does. Sucrase's react-hot-loader transform is really just a few extra details for the react-hot-loader project to work. As far as I understand, HMR is a webpack feature that's also a bit tied into webpack's module resolution system, so my guess is your best bet is to try to just use webpack.

Have you seen https://github.com/webpack/hot-node-example ? Seems like that might help, though it's pretty old.

I also saw your previous attempt at https://github.com/konsumer/ink-demo/commit/367b96f07d6380bfc18f3b863dbd9b9cee4f2389#diff-1197ba97f47af352583345d4fad71299 . One thought is that IIRC to really get hot reloading to work, you need to have the old and new module up at the same time so you can hand off the state. But I'm certainly no expert in that stuff.

**Comment by konsumer** [comment]
> Have you seen https://github.com/webpack/hot-node-example ? Seems like that might help, though it's pretty old.

Yep, I saw that, but couldn't get it working with ink. I also tried some of the tactics in [react-blessed-hot-motion](https://github.com/gaearon/react-blessed-hot-motion), which I also couldn't get to work (and don't seem to work with react-blessed anymore.)

> One thought is that IIRC to really get hot reloading to work, you need to have the old and new module up at the same time so you can hand off the state. But I'm certainly no expert in that stuff.

This is a great pointer to what might be the problem. Thanks!

I will keep playing with it. Sucrase is so much faster, I'd rather get it working, anyway. I will update if I come up with anything.

Here is the ink issue: [vadimdemedes/ink#159](https://github.com/vadimdemedes/ink/issues/159)

**Comment by alangpierce** [comment]
Another thought that might give you a similarly-nice dev cycle: in a past project that I worked on, I was able to iterate pretty well by storing 100% of app state in redux (never in component state) and saving all redux state to localstorage after every action. Then, on app startup, it would load localstorage into redux if necessary. If you design your ink app like that (maybe a JSON file instead of localstorage), then you could have a watcher to restart the CLI whenever any file changes and it'll start off from the latest state.

Of course, a nice thing about react-hot-loader is you can just use plain component state without having to think about that stuff.

**Comment by konsumer** [comment]
In my case, I think I'm ok with state, at least for the simple demo I am working with, which has pretty minimal state (a counter hook, to demonstrate that dynamic value & hooks are working.)

Your comments have lead me to the idea that I probly need to get webpack working even though it's not a web project, or at least something that holds the old & new source and swaps it out in the same manner. react-native does it their own way, I think, in the builder. I will investigate when I have more time.

**Comment by alangpierce** [comment]
> I probly need to get webpack working even though it's not a web project, or at least something that holds the old & new source and swaps it out in the same manner

FWIW I think it should be possible to require a module, invalidate the module cache, then require it again as a way to get the old and new state.

**Comment by konsumer** [comment]
> FWIW I think it should be possible to require a module, invalidate the module cache, then require it again as a way to get the old and new state.

I tried this at first, but I don't think I was holding the 2 versions of code and swapping them, correctly. I will explore this more.
